modules/models/behavior/motion_primitives/macro_actions.cpp

```diff
@@ -23,23 +28,77 @@ BehaviorMPMacroActions::AddMotionPrimitive(const primitives::PrimitivePtr& primi
 Trajectory BehaviorMPMacroActions::Plan(
     float delta_time, const world::ObservedWorld& observed_world) {
   SetBehaviorStatus(BehaviorStatus::VALID);
-  // TODO: move this to Predicate
   const float dt = integration_time_delta_;
   const int num_trajectory_points =
       static_cast<int>(std::ceil(delta_time / dt)) + 1;
 
   Trajectory traj(num_trajectory_points,
                   static_cast<int>(StateDefinition::MIN_STATE_SIZE));
 
-  traj = motion_primitives_.at(
-    boost::get<DiscreteAction>(active_motion_))->Plan(delta_time, observed_world);
-  // traj = 
 
   // SetLastAction(Action(DiscreteAction(active_motion_)));
+  AdjacentLaneCorridors adjacent_corridors = GetCorridors(observed_world);
+  // There must be at least one primitive that is always available!
+  if(valid_primitives_.empty()) {
+    GetNumMotionPrimitivesByCorridors(observed_world, adjacent_corridors);
+    LOG_IF(ERROR, valid_primitives_.empty()) << "No motion primitive available! At least one primitive must be available at all times!";
+  }
+  const auto& selected_mp = motion_primitives_.at(valid_primitives_.at(boost::get<DiscreteAction>(active_motion_)));
+  target_corridor_ = selected_mp->SelectTargetCorridor(observed_world, adjacent_corridors);
+  traj = selected_mp->Plan(delta_time, observed_world, target_corridor_);
 
   this->SetLastTrajectory(traj);
   return traj;
 }
+const std::vector<primitives::PrimitivePtr>&
+BehaviorMPMacroActions::GetMotionPrimitives() const {
+  return motion_primitives_;
+}
+BehaviorMotionPrimitives::MotionIdx
+BehaviorMPMacroActions::GetNumMotionPrimitives(
+    const ObservedWorldPtr& observed_world) {
+  return GetNumMotionPrimitivesByCorridors(*observed_world,
+                                           GetCorridors(*observed_world));
+}
+AdjacentLaneCorridors BehaviorMPMacroActions::GetCorridors(
+    const ObservedWorld& observed_world) {
+  if(!target_corridor_) {
+    target_corridor_ = observed_world.GetLaneCorridor();
+  }
+  AdjacentLaneCorridors adjacent_corridors;
+  auto ego_pos = observed_world.CurrentEgoPosition();
+  auto point_on_target_line =
+      GetNearestPoint(target_corridor_->GetCenterLine(), ego_pos);
+  auto road_corridor = observed_world.GetRoadCorridor();
+  std::tie(adjacent_corridors.left, adjacent_corridors.right) =
+      road_corridor->GetLeftRightLaneCorridor(point_on_target_line);
+  adjacent_corridors.current = target_corridor_;
+  return adjacent_corridors;
+}
+BehaviorMotionPrimitives::MotionIdx
+BehaviorMPMacroActions::GetNumMotionPrimitivesByCorridors(
+    const ObservedWorld& observed_world,
+    const AdjacentLaneCorridors& adjacent_corridors) {
+  MotionIdx i = 0;
```

Comment:
is i used?

Reply:
Yes here https://github.com/bark-simulator/bark/blob/0eefd3587137099d0a21ddf8a3bf8b4b9b4b19df/modules/models/behavior/motion_primitives/macro_actions.cpp#L86